### PR TITLE
Add persistent+TMA version of Triton mm and addmm

### PR DESCRIPTION
**Pull request by aakhundov**
This PR adds persistent+TMA versions (Triton template + the corresponding infra) for the `tuned_mm` and `tuned_addmm` lowerings. The persistent+TMA choices are added to the GEMM autotuning if (checked by the `use_triton_tma_template` helper):

1. The min. hardware and Triton version requirements are met for the TMA support.

2. The GEMM inputs are compatible with the Triton TMA API (i.e., 16-byte aligned and contiguous).

3. The `config.triton.enable_persistent_tma_matmul` is set to `True`.

Additional notes:

1. As added in this PR, the TMA uses are not compatible with prolog / epilogue fusion. To this end, in the new Triton template we currently support: TMA-based loads of A/B, but no prologue fusion; epilogue fusion, but no TMA-based stores of C. TMA + fusion compatibility can be added as a follow-up.

2. The current Triton TMA API (`experimental_device_tensormap_create2d`) does not support strides. Due to this, we limit the applicability of the new Triton template to the cases where the inputs are contiguous.

3. The transposed layouts of A and / or B are supported by passing the constexpr flags to the kernel and adjusting the ordering of the block sizes accordingly in the kernel code (this should have no effect on the kernel perf, as decided at the Triton compilation time).

4. After the next Triton pin update, we can switch to the tensor descriptor API (landed recently in https://github.com/triton-lang/triton/pull/5290) in the new Triton template, which should allow lifting 2 and 3 above.

5. The configs for the new Triton template in `persistent_mm_kernel_configs` are preliminary. We should do more perf exploration and possibly augment the config in a follow-up.

6. This PR is rebased onto and unifies with two related PRs landed previously: https://github.com/pytorch/pytorch/pull/142045 (some infra unification with the persistent+TMA template for _scaled_mm) and https://github.com/pytorch/pytorch/pull/134532 (add possibility to disable prolog fusion for selected choices).

7. The current Triton TMA API only supports 1D and 2D descriptors (even after https://github.com/triton-lang/triton/pull/5290, see [here](https://github.com/triton-lang/triton/blob/9829ce87ccb333a2b264b3a80b39a534bfa865ac/python/triton/language/core.py#L1957)). For now, this blocks adding persistent+TMA template for `torch.bmm`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142101

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/scheduler.py [hunk 1]
```diff
@@ -2645,6 +2645,18 @@ def log_fusion(ms_fused: float, ms1: float, ms2: float) -> None:
...
+                # supports all allowed prologue inputs. If not, we skip this choice in
+                # the fusion benchmark.
+                # TODO: Remove this check after all Triton templates support prologue fusion.
+                # Currently, persistent+TMA Triton template does not due to the TMA-based loads.
+                if (
+                    not epilogue_fusion
+                    and hasattr(choice, "allowed_prologue_inps")
+                    and choice.allowed_prologue_inps != multi_node.allowed_prologue_inps
+                ):
+                    continue
```
Comment (aakhundov) [inline torch/_inductor/scheduler.py:2658]: cc @eellison: this is to selectively skip choices not supporting prologue fusion (like currently the choices from the persistent+TMA template).

File: torch/_inductor/kernel/mm_common.py [hunk 2]
```diff
@@ -336,6 +336,18 @@ def filtered_configs(
...
+    {"config": (128, 128, 64, 3, 8), "cond": True},
+    {"config": (128, 128, 128, 3, 8), "cond": True},
+    {"config": (128, 128, 128, 4, 8), "cond": True},
+    {"config": (128, 128, 128, 4, 4), "cond": True},
+    {"config": (128, 128, 128, 3, 4), "cond": True},
+    {"config": (128, 128, 128, 5, 4), "cond": True},
+    {"config": (128, 128, 128, 5, 8), "cond": True},
+    {"config": (128, 128, 128, 6, 8), "cond": True},
+    {"config": (128, 128, 64, 4, 8), "cond": True},
+]
```
Comment (aakhundov) [inline torch/_inductor/kernel/mm_common.py:349]: cc @drisspg: separated these configs used in `_scaled_mm` persistent+TMA template-based lowering, as ~half of them OOMs on SMEM for 2-byte dtypes. Kept the ones that don't in `persistent_mm_kernel_configs`.

File: test/inductor/test_fp8.py [hunk 3]
```diff
@@ -655,7 +655,7 @@ def linear(x_fp8, x_inverse_scale, w_t_fp8, w_inverse_scale, bias):
             w_inverse_scale,
             bias,
         )
-        with config.patch({"triton.enable_persistent_tma_matmul": True}):
```
Comment (drisspg) [inline test/inductor/test_fp8.py:658]: wow, bad mistake on my end thanks for fixing!

File: torch/_inductor/kernel/mm.py [hunk 4]
```diff
@@ -128,6 +133,112 @@
...
+    M = {{size("A", 0)}}
+    N = {{size("B", 1)}}
+    K = {{size("A", 1)}}
+    if M * N == 0:
+        # early exit due to zero-size input(s)
+        return
+
+    # based on triton.ops.matmul
+    start_pid = tl.program_id(0)
+    grid_m = (M + BLOCK_M - 1) // BLOCK_M
```
Comment (drisspg) [inline torch/_inductor/kernel/mm.py:150]: nit: why not cdiv for these too?

File: torch/_inductor/kernel/mm_scaled.py [hunk 5]
```diff
@@ -566,8 +552,9 @@ def tuned_scaled_mm(
                     choices,
                     input_nodes=input_nodes,
                     layout=layout,
-                    workspace_arg=get_workspace_arg(
-                        kwargs["NUM_SMS"], mat_a.get_device()
+                    workspace_arg=get_tma_workspace_arg(
+                        num_tma_descriptors=3,
```
Comment (drisspg) [inline torch/_inductor/kernel/mm_scaled.py:556]: we actually only need 2, if you care to update as well, my top of stack had the C stores but that was buggy anyways

**Review by drisspg — APPROVED** [review]
Looks great! will let Elias comment on the prologue stuff

File: torch/_inductor/kernel/mm_scaled.py (hunk 5 above)
Comment (aakhundov) [reply torch/_inductor/kernel/mm_scaled.py:556]: Thanks for catching this! Helped me find a nasty bug in the new template code, too.

File: torch/_inductor/kernel/mm.py [hunk 6]
```diff
@@ -128,6 +133,110 @@
...
+
+        rk = ki * BLOCK_K
+
+        a = tl._experimental_descriptor_load(
+            a_desc_ptr,
+            [rm, rk] if A_ROW_MAJOR else [rk, rm],
+            [BLOCK_M, BLOCK_K] if A_ROW_MAJOR else [BLOCK_K, BLOCK_M],
+            A.dtype.element_ty,
+        )
+        b = tl._experimental_descriptor_load(
```
Comment (eellison) [inline torch/_inductor/kernel/mm.py:212]: i dont see any mask here. i guess it doesnt support k not divisible by k block ?

**Review by eellison — APPROVED** [review]
looks great !

File: torch/_inductor/select_algorithm.py [hunk 7]
```diff
@@ -1632,6 +1633,25 @@ def __call__(
         # TODO(nmacchioni): remove once CI tests are fixed
         choices = [choice for choice in choices if choice is not None]
 
+        if config.test_configs.autotune_choice_name_regex is not None:
+            choices = [
+                c
+                for c in choices
+                if re.search(
+                    config.test_configs.autotune_choice_name_regex,
```
Comment (eellison) [inline torch/_inductor/select_algorithm.py:1641]: nit: it is it worth functools.lru_caching re.compile() and using that ?

File: torch/_inductor/kernel/mm.py [hunk 8]
```diff
@@ -128,6 +133,110 @@
...
+            # rematerialize rm and rn to save registers
+            rcm = rm + tl.arange(0, BLOCK_M)
+            rcn = rn + tl.arange(0, BLOCK_N)
+            idx_m = rcm[:, None]
+            idx_n = rcn[None, :]
+            mask = (idx_m < M) & (idx_n < N)
+
+            # inductor generates a suffix
+            {{store_output(("idx_m", "idx_n"), "acc", "mask", indent_width=12)}}
+            acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=ACC_TYPE)
```
[the remark this answers is not in the document]
Comment (drisspg) [reply torch/_inductor/kernel/mm.py:236]: yeah I we should

File: torch/_inductor/kernel/mm.py (hunk 6 above)
Comment (aakhundov) [reply torch/_inductor/kernel/mm.py:212]: TMA does support k not divisible by k block (given that `k * dtype.itemsize % 16 == 0`). Masking happens in the HW doing TMA, with the OOB values set to zero.

File: torch/_inductor/select_algorithm.py (hunk 7 above)
Comment (aakhundov) [reply torch/_inductor/select_algorithm.py:1641]: I imagine these regexes being used exclusively for testing (as in: not in prod / with real models) and being relatively simple (prob a substring or a few separated by |). So not sure how much it's worth precompiling them. Happy to add if you feel it's worth, though.

File: torch/_inductor/kernel/mm.py [hunk 9]
```diff
@@ -128,6 +133,110 @@
...
+        if B_PROLOGUE_CAST_TYPE is not None:
+            b = b.to(B_PROLOGUE_CAST_TYPE)
+        acc += tl.dot(
+            a if A_ROW_MAJOR else a.T,
+            b if B_ROW_MAJOR else b.T,
+            allow_tf32=ALLOW_TF32,
+        )
+
+        if ki == k_tiles - 1:
+            # rematerialize rm and rn to save registers
```
Comment (kimishpatel) [inline torch/_inductor/kernel/mm.py:227]: how do you determine if not doing this results in spill/fills?

**Pull request closed without merging**